File: js/src/attestation.ts
/**
 * Uniplex Attestation - Signed proof of authorization.
 *
 * An Attestation is a cryptographically signed record of an authorization decision.
 * It provides non-repudiable evidence that a specific check was performed.
 */

import * as crypto from './crypto.js';
import { GateRequest, GateDecision } from './gate.js';

// ============================================================================
// Types
// ============================================================================

export interface AttestationData {
  uni_version: string;
  attestation_id: string;
  request_id: string;
  passport_id: string | null;
  agent_id: string | null;
  action: string;
  target?: string | null;
  decision: 'allow' | 'deny';
  reason_code?: string | null;
  created_at: string;
  gate_id?: string | null;
  public_key?: string;
  signature?: string;
}

// ============================================================================
// Helper Functions
// ============================================================================

function generateAttestationId(): string {
  const chars = 'abcdef0123456789';
  let id = 'att_';
  for (let i = 0; i < 12; i++) {
    id += chars[Math.floor(Math.random() * chars.length)];
  }
  return id;
}

// ============================================================================
// Attestation Class
// ============================================================================

export class Attestation {
  readonly uniVersion: string;
  readonly attestationId: string;
  readonly requestId: string;
  readonly passportId: string | null;
  readonly agentId: string | null;
  readonly action: string;
  readonly target?: string | null;
  readonly decision: 'allow' | 'deny';
  readonly reasonCode?: string | null;
  readonly createdAt: string;
  readonly gateId?: string | null;
  publicKey?: string;
  signature?: string;

  private _privateKey?: Uint8Array;

  constructor(data: AttestationData, privateKey?: Uint8Array) {
    this.uniVersion = data.uni_version;
    this.attestationId = data.attestation_id;
    this.requestId = data.request_id;
    this.passportId = data.passport_id;
    this.agentId = data.agent_id;
    this.action = data.action;
    this.target = data.target;
    this.decision = data.decision;
    this.reasonCode = data.reason_code;
    this.createdAt = data.created_at;
    this.gateId = data.gate_id;
    this.publicKey = data.public_key;
    this.signature = data.signature;
    this._privateKey = privateKey;
  }

  /**
   * Create an attestation from a gate decision.
   */
  static async fromDecision(
    request: GateRequest,
    decision: GateDecision,
    options: {
      gateId?: string;
    } = {}
  ): Promise<Attestation> {
    const { privateKey, publicKey } = await crypto.generateKeypair();

    const attestation = new Attestation(
      {
        uni_version: '2026-01-25',
        attestation_id: generateAttestationId(),
        request_id: request.requestId,
        passport_id: decision.passportId ?? null,
        agent_id: decision.agentId ?? null,
        action: request.action,
        target: request.target,
        decision: decision.decision,
        reason_code: decision.reasonCode ?? null,
        created_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
        gate_id: options.gateId ?? null,
        public_key: crypto.encodeBase64(publicKey),
      },
      privateKey
    );

    await attestation.sign();
    return attestation;
  }

  /**
   * Sign the attestation.
   */
  async sign(): Promise<void> {
    if (!this._privateKey) {
      throw new Error('No private key available for signing');
    }

    const dataToSign = this.toDict();
    delete dataToSign.signature;

    const message = crypto.objectToBytes(dataToSign);
    const sig = await crypto.sign(this._privateKey, message);
    this.signature = crypto.encodeBase64(sig);
  }

  /**
   * Verify the attestation signature.
   */
  async verifySignature(): Promise<boolean> {
    if (!this.signature || !this.publicKey) {
      return false;
    }

    try {
      const dataToVerify = this.toDict();
      delete dataToVerify.signature;

      const message = crypto.objectToBytes(dataToVerify);
      const signature = crypto.decodeBase64(this.signature);
      const publicKey = crypto.decodeBase64(this.publicKey);

      return await crypto.verify(publicKey, signature, message);
    } catch {
      return false;
    }
  }

  /**
   * Convert to plain object.
   */
  toDict(): AttestationData {
    return {
      uni_version: this.uniVersion,
      attestation_id: this.attestationId,
      request_id: this.requestId,
      passport_id: this.passportId,
      agent_id: this.agentId,
      action: this.action,
      target: this.target,
      decision: this.decision,
      reason_code: this.reasonCode,
      created_at: this.createdAt,
      gate_id: this.gateId,
      public_key: this.publicKey,
      signature: this.signature,
    };
  }

  /**
   * Convert to JSON string.
   */
  toJson(): string {
    return JSON.stringify(this.toDict(), null, 2);
  }

  /**
   * Create from plain object.
   */
  static fromDict(data: AttestationData): Attestation {
    return new Attestation(data);
  }

  /**
   * Create from JSON string.
   */
  static fromJson(json: string): Attestation {
    return Attestation.fromDict(JSON.parse(json));
  }
}

File: uniplex/attestation.py
"""
Uniplex Attestation - Proof of authorization decisions.

An Attestation is a signed record of a Gate decision. It provides:
- Proof that an authorization check occurred
- What was checked (passport, action, target)
- What was decided (allow/deny)
- When it happened
- Cryptographic proof of integrity

Attestations can be stored, transmitted, and verified later.
"""

import json
import uuid
from datetime import datetime, timezone
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from . import crypto
from .gate import GateDecision, GateRequest


class Attestation(BaseModel):
    """
    Uniplex Attestation - Signed proof of an authorization decision.
    
    Attestations provide an audit trail and can be verified by third parties
    to confirm that an authorization check was performed.
    """
    
    uni_version: str = "2026-01-25"
    attestation_id: str = Field(default_factory=lambda: f"att_{uuid.uuid4().hex[:16]}")
    
    # What was attested
    request_id: str
    passport_id: Optional[str] = None
    agent_id: Optional[str] = None
    action: str
    target: Optional[str] = None
    
    # The decision
    decision: str  # "allow" or "deny"
    reason_code: Optional[str] = None
    
    # Timing
    created_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    )
    
    # Gate identity
    gate_id: Optional[str] = None
    
    # Cryptographic proof
    public_key: Optional[str] = None  # Base64 encoded
    signature: Optional[str] = None  # Base64 encoded
    
    # Internal - not serialized
    _private_key: Optional[bytes] = None
    
    model_config = ConfigDict(extra="allow")
    
    @classmethod
    def from_decision(
        cls,
        request: GateRequest,
        decision: GateDecision,
        gate_id: Optional[str] = None,
        sign: bool = True,
    ) -> "Attestation":
        """
        Create an attestation from a gate decision.
        
        Args:
            request: The original authorization request
            decision: The gate's decision
            gate_id: Optional identifier for the gate
            sign: Whether to sign the attestation (default True)
        
        Returns:
            A new Attestation, optionally signed
        """
        attestation = cls(
            request_id=request.request_id,
            passport_id=decision.passport_id,
            agent_id=decision.agent_id,
            action=request.action,
            target=request.target,
            decision=decision.decision,
            reason_code=decision.reason_code,
            gate_id=gate_id,
        )
        
        if sign:
            attestation.sign_self()
        
        return attestation
    
    def sign_self(self) -> None:
        """
        Generate a new keypair and sign the attestation.
        
        This is useful for quick attestation generation where the
        gate doesn't have persistent identity.
        """
        private_key, public_key = crypto.generate_keypair()
        self._private_key = crypto.private_key_to_bytes(private_key)
        self.public_key = crypto.encode_base64(crypto.public_key_to_bytes(public_key))
        self._sign()
    
    def sign_with_key(self, private_key_bytes: bytes) -> None:
        """
        Sign the attestation with an existing private key.
        
        Args:
            private_key_bytes: Raw bytes of the Ed25519 private key
        """
        private_key = crypto.private_key_from_bytes(private_key_bytes)
        public_key = private_key.public_key()
        
        self._private_key = private_key_bytes
        self.public_key = crypto.encode_base64(crypto.public_key_to_bytes(public_key))
        self._sign()
    
    def _sign(self) -> None:
        """Sign the attestation with the stored private key."""
        if self._private_key is None:
            raise ValueError("No private key available for signing")
        
        private_key = crypto.private_key_from_bytes(self._private_key)
        
        # Get canonical JSON of attestation (without signature)
        data = self.to_signable_dict()
        message = crypto.canonical_json(data)
        
        # Sign
        sig = crypto.sign(private_key, message)
        self.signature = crypto.encode_base64(sig)
    
    def to_signable_dict(self) -> dict:
        """Get attestation data for signing (excludes signature field)."""
        data = self.model_dump(exclude={"signature"}, exclude_none=True)
        # Remove any private fields
        return {k: v for k, v in data.items() if not k.startswith("_")}
    
    def to_dict(self) -> dict:
        """Export attestation as dictionary."""
        data = self.model_dump(exclude_none=True)
        return {k: v for k, v in data.items() if not k.startswith("_")}
    
    def to_json(self, indent: Optional[int] = 2) -> str:
        """Export attestation as JSON string."""
        return json.dumps(self.to_dict(), indent=indent)
    
    def verify_signature(self) -> bool:
        """
        Verify the attestation's signature.
        
        Returns True if signature is valid, False otherwise.
        """
        if not self.signature or not self.public_key:
            return False
        
        try:
            public_key_bytes = crypto.decode_base64(self.public_key)
            signature_bytes = crypto.decode_base64(self.signature)
            public_key = crypto.public_key_from_bytes(public_key_bytes)
            
            data = self.to_signable_dict()
            message = crypto.canonical_json(data)
            
            return crypto.verify(public_key, signature_bytes, message)
        except Exception:
            return False
    
    @classmethod
    def from_json(cls, json_str: str) -> "Attestation":
        """Parse attestation from JSON string."""
        data = json.loads(json_str)
        return cls.model_validate(data)
    
    @classmethod
    def from_dict(cls, data: dict) -> "Attestation":
        """Create attestation from dictionary."""
        return cls.model_validate(data)
    
    @property
    def allowed(self) -> bool:
        """Check if this attestation records an allow decision."""
        return self.decision == "allow"
    
    @property
    def denied(self) -> bool:
        """Check if this attestation records a deny decision."""
        return self.decision == "deny"
